validator/src/journal/block_store.rs

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::HashMap;
+
+use block::Block;
+
+pub enum BlockStoreError {
+    Error(String),
+}
+
+pub trait BlockStore {
+    fn get<'a>(&'a self, block_ids: Vec<String>) -> Box<Iterator<Item = &'a Block> + 'a>;
+
+    fn put(&mut self, blocks: Vec<Block>) -> Result<(), BlockStoreError>;
+
+    fn iter(&self) -> Box<Iterator<Item = Block>>;
+}
+
+#[derive(Default)]
+pub struct InMemoryBlockStore {
+    block_by_block_id: HashMap<String, Block>,
+}
+
+impl InMemoryBlockStore {
+    pub fn new() -> Self {
+        InMemoryBlockStore::default()
+    }
+
+    fn get_block_by_block_id(&self, block_id: &str) -> Option<&Block> {
+        self.block_by_block_id.get(block_id)
+    }
+}
+
+impl BlockStore for InMemoryBlockStore {
+    fn get<'a>(&'a self, block_ids: Vec<String>) -> Box<Iterator<Item = &'a Block> + 'a> {
+        let iterator: GetBlockIterator = GetBlockIterator::new(self, block_ids);
+
+        Box::new(iterator)
+    }
+
+    fn put(&mut self, blocks: Vec<Block>) -> Result<(), BlockStoreError> {
+        blocks.into_iter().for_each(|block| {
+            self.block_by_block_id
+                .insert(block.header_signature.clone(), block);
+        });
+        Ok(())
+    }
+
+    fn iter(&self) -> Box<Iterator<Item = Block>> {
+        unimplemented!()
+    }
+}
+
+struct GetBlockIterator<'a> {
+    blockstore: &'a InMemoryBlockStore,
```

Comment:
This probably shouldn't be tied to `InMemoryBlockStore`, or should be called `InMemoryBlockStoreIterator`

Reply:
I can probably do &'a BlockStore, because all we care about is that it is a BlockStore.